docs/src/main/asciidoc/application-data-caching-guide.adoc

```diff
@@ -0,0 +1,433 @@
+If several threads try to retrieve a cache value from the same missing key concurrently, the method will only be invoked once.
+The first calling thread will trigger the method invocation while the subsequent calling threads will wait for the end of the invocation to get the cached result.
+The `lockTimeout` parameter can be used to interrupt the lock after a given delay.
+See the parameter Javadoc for more details.
+
+This annotation can neither be used on a method returning `void` nor in combination with another Quarkus method caching annotation.
+
+=== @CacheInvalidate
+
+Removes an entry from the cache.
+
+When a method annotated with `@CacheInvalidate` is invoked, Quarkus will compute a cache key from the method arguments (or generate a key if the method has no arguments)
+and use it to try to remove an existing entry from the cache.
+The key computation is done from all the method arguments if none of them is annotated with `@CacheKey`, or all the arguments annotated with `@CacheKey` otherwise.
+If the key does not identify any cache entry, nothing will happen.
+
+This annotation can't be used in combination with another Quarkus method caching annotation.
+
+[TIP]
+====
+If the `@CacheResult` or `@CacheInvalidate` annotations are used on a method with no parameters, a unique default cache key derived from the cache name will be generated and used.
+====
+
+=== @CacheInvalidateAll
+
+When a method annotated with `@CacheInvalidateAll` is invoked, Quarkus will remove all entries from the cache.
+
+This annotation can't be used in combination with another Quarkus method caching annotation.
+
+=== @CacheKey
+
+When a method argument is annotated with `@CacheKey`, it is identified as a part of the cache key during an invocation of a
+method annotated with `@CacheResult` or `@CacheInvalidate`.
+
+This annotation is optional and should only be used when some of the method arguments are NOT part of the cache key.
+
+== Configuring the underlying caching provider
+
+For now, https://github.com/ben-manes/caffeine[Caffeine] is the only available caching provider in Quarkus.
+We plan to add more providers soon.
+
+=== Caffeine configuration properties
+
+Caffeine is a high performance, near optimal caching library.
+
+Each of the Caffeine caches backing up the Quarkus application data caching extension can be configured using the following
+properties in the `application.properties` file.
+
+[TIP]
+====
+You need to replace `cache-name` in all of the following properties with the real name of the cache you want to configure.
+====
+
+include::{generated-dir}/config/quarkus-cache-config-group-cache-config-caffeine-config.adoc[opts=optional, leveloffset=+1]
+
+Here's what your cache configuration could look like:
+
+[source,properties]
+----
+quarkus.cache.caffeine."foo".initial-capacity=10 <1>
+quarkus.cache.caffeine."foo".maximum-size=20 <1>
+quarkus.cache.caffeine."foo".expire-after-write <1>
+quarkus.cache.caffeine."bar".maximum-size=1000 <2>
```

Comment:
So I was surprised by the forced `quarkus.cache.caffeine.*` vs `quarkus.cache.*` but I guess it's ok to start like that. We can go and move them to a generic `quarkus.cache` later while keeping backward compatibility.

Reply:
The `caffeine` segment of the key was introduced because the initial plan was to have multiple caching providers in this extension. As you wrote, it doesn't hurt to leave the key that way for now and introduce the generic key later.